File: kudos-ao3-reader/Features/Search/FandomCatalogCache.swift
import Foundation

/// Local-first, on-disk cache of each media category's full fandom list, so the
/// heavy `/media/<category>/fandoms` pages (thousands of fandoms each) aren't
/// re-scraped every time the user opens Browse. Stored as a single JSON file in the
/// evictable metadata cache directory; each entry carries its fetch date so stale
/// ones can be refreshed in the background while the cached copy shows immediately.
struct FandomCatalogCache {
    struct Entry: Codable {
        var fandoms: [AO3Fandom]
        var fetchedAt: Date
    }

    /// Fandom counts drift slowly; refresh a category at most about once a week.
    static let maxAge: TimeInterval = 7 * 24 * 60 * 60

    private let fileURL: URL

    init(fileURL: URL = FandomCatalogCache.defaultFileURL) {
        self.fileURL = fileURL
    }

    static var defaultFileURL: URL {
        Storage.metadataCacheDirectory.appendingPathComponent("fandom-catalog.json")
    }

    /// Reads the cached entries (keyed by category id). Empty if absent/unreadable.
    /// Safe to call off the main actor — it only touches the filesystem.
    func load() -> [String: Entry] {
        guard let data = try? Data(contentsOf: fileURL),
              let decoded = try? JSONDecoder().decode([String: Entry].self, from: data)
        else { return [:] }
        return decoded
    }

    /// Atomically writes the entries. Best-effort: a failed write just means the
    /// next launch re-scrapes. Safe to call off the main actor.
    func save(_ entries: [String: Entry]) {
        guard let data = try? JSONEncoder().encode(entries) else { return }
        try? data.write(to: fileURL, options: .atomic)
    }

    /// Removes the cache file (the Privacy & Local Data "Clear" action). It rebuilds
    /// the next time the user opens Browse.
    func clear() {
        try? FileManager.default.removeItem(at: fileURL)
    }

    /// Whether a category needs (re)fetching: missing, or older than `maxAge`.
    static func isStale(_ entry: Entry?, now: Date = Date()) -> Bool {
        guard let entry else { return true }
        return now.timeIntervalSince(entry.fetchedAt) > maxAge
    }
}

File: android/app/src/main/java/io/github/cidy02/kudos/network/ao3/browse/FandomCatalogCache.kt
package io.github.cidy02.kudos.network.ao3.browse

import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.time.Duration
import java.time.Instant
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Local-first, on-disk cache of each media category's full fandom list, so the
 * heavy `/media/<category>/fandoms` pages (thousands of fandoms each) aren't
 * re-scraped every time Browse opens. Mirrors iOS `FandomCatalogCache`: a
 * single JSON file in the evictable cache directory, keyed by category name,
 * each entry carrying its fetch time so staleness can be checked without a
 * network call.
 */
class FandomCatalogCache(private val cacheDir: Path) {
    @Serializable
    data class Entry(val fandoms: List<AO3Fandom>, val fetchedAtEpochMillis: Long)

    private val json = Json { ignoreUnknownKeys = true }
    private val filePath: Path get() = cacheDir.resolve(FILE_NAME)

    /** Reads cached entries keyed by category name. Empty if absent/unreadable. */
    suspend fun load(): Map<String, Entry> = withContext(Dispatchers.IO) {
        if (!Files.isRegularFile(filePath)) return@withContext emptyMap()
        runCatching {
            json.decodeFromString<Map<String, Entry>>(Files.readString(filePath))
        }.getOrDefault(emptyMap())
    }

    /**
     * Atomically writes the entries (temp file + move, mirrors
     * [io.github.cidy02.kudos.files.FontFileStore]). Best-effort: a failed
     * write just means the next launch re-scrapes.
     */
    suspend fun save(entries: Map<String, Entry>) = withContext(Dispatchers.IO) {
        runCatching {
            Files.createDirectories(cacheDir)
            val temp = Files.createTempFile(cacheDir, ".fandom-catalog-", ".json.tmp")
            try {
                Files.writeString(temp, json.encodeToString(entries))
                try {
                    Files.move(
                        temp,
                        filePath,
                        StandardCopyOption.REPLACE_EXISTING,
                        StandardCopyOption.ATOMIC_MOVE
                    )
                } catch (_: Exception) {
                    Files.move(temp, filePath, StandardCopyOption.REPLACE_EXISTING)
                }
            } finally {
                Files.deleteIfExists(temp)
            }
        }
        Unit
    }

    /** Removes the cache file (Privacy & Local Data "Clear Browse Cache"). Rebuilds next open. */
    suspend fun clear() = withContext(Dispatchers.IO) {
        runCatching { Files.deleteIfExists(filePath) }
        Unit
    }

    companion object {
        private const val FILE_NAME = "fandom-catalog.json"

        /** Fandom counts drift slowly; refresh a category at most about once a week. */
        val MaxAge: Duration = Duration.ofDays(7)

        fun isStale(entry: Entry?, now: Instant = Instant.now()): Boolean {
            if (entry == null) return true
            return Duration.between(Instant.ofEpochMilli(entry.fetchedAtEpochMillis), now) > MaxAge
        }
    }
}
